File: LeetCode/Java/49-hash-table-字母异位词分组.java
/*
 * @lc app=leetcode.cn id=49 lang=java
 *
 * [49] 字母异位词分组
 */

import java.util.*;

// @lc code=start
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        if (strs.length == 0)
            return new ArrayList();
        var res = new HashMap<String, List>();

        for (String s : strs) {
            var ca = s.toCharArray();
            Arrays.sort(ca);
            var key = String.valueOf(ca);
            if (!res.containsKey(key))
                res.put(key, new ArrayList());
            res.get(key).add(s);
        }

        return new ArrayList(res.values());
    }
}
// @lc code=end

File: LeetCode/C/49-hash-table-字母异位词分组.cpp
/*
 * @lc app=leetcode.cn id=49 lang=cpp
 *
 * [49] 字母异位词分组
 */
#include "main.h"

// @lc code=start
class Solution {
  public:
    vector<vector<string>> groupAnagrams(vector<string> &strs) {
        unordered_map<string, vector<string>> mp;
        for (string str : strs) {
            int count[26] = {0};
            for (char c : str)
                ++count[c - 'a'];
            string key;
            for (int i = 0; i < 26; i++)
                key += ('#' + to_string(count[i]));
            mp[key].push_back(str);
        }
        vector<vector<string>> res;
        for (auto it = mp.begin(); it != mp.end(); it++)
            res.push_back((*it).second);
        return res;
    }
};
// @lc code=end
